File: Astro-TFG/public/scripts/6-patrones/prueba-1/version-1/javascript/main.js
let worker = null;
let workerTime = 0;

function initializeWorker() {
    if (!worker) {
        worker = new Worker(
            window.location.origin +
            window.document.body.dataset.jsPath +
            "worker.js",
            { type: "module" }
        );
    }
}

async function _runJsBenchmark() {
    try {
        window.clearCell("javascript-output");
        window.showExecutionLoader();

        const startWorkerTime = performance.now();
        await initializeWorker();
        workerTime = performance.now() - startWorkerTime;

        const repetitions = parseInt(
            document.getElementById("num-repetitions-javascript").value,
            10
        );

        const id = `js-${Date.now()}`;

        const resultJson = await new Promise((resolve, reject) => {
            function onMessage(e) {
                if (e.data.id !== id) return;
                worker.removeEventListener("message", onMessage);
                if (e.data.error) {
                    reject(new Error(e.data.error));
                } else {
                    resolve(e.data.json);
                }
            }
            worker.addEventListener("message", onMessage);
            worker.postMessage({
                id,
                type: "do_analisis",
                repetitions,
            });
        });

        const r = JSON.parse(resultJson);
        displayResult(r);
    } catch (err) {
        console.error("Benchmark error:", err);
        document.getElementById(
            "javascript-output"
        ).textContent = `Error: ${err.message}`;
    } finally {
        window.hideExecutionLoader();
    }
}

function createDiv() {
    const div = document.createElement("div");
    return div;
}

function displayResult(r) {
    const output = document.getElementById("javascript-output");

    const workerDiv = createDiv();
    workerDiv.textContent = `Worker init time: ${workerTime.toFixed(2)} ms`;
    output.appendChild(workerDiv);

    const trainingDiv = createDiv();
    trainingDiv.textContent = `Training time: ${r.training_time_ms.toFixed(2)} ms`;
    output.appendChild(trainingDiv);

    const inferenceDiv = createDiv();
    inferenceDiv.textContent = `Inference time: ${r.inference_time_ms.toFixed(2)} ms`;
    output.appendChild(inferenceDiv);

    const accuracyDiv = createDiv();
    accuracyDiv.textContent = `Accuracy: ${r.accuracy.toFixed(2)} %`;
    output.appendChild(accuracyDiv);

    const modelSizeDiv = createDiv();
    modelSizeDiv.textContent = `Model size: ${r.model_size_mb.toFixed(2)} MB`;
    output.appendChild(modelSizeDiv);

    const totalDiv = createDiv();
    totalDiv.textContent = `Total ET: ${r.overall_time_ms.toFixed(2)} ms`;
    output.appendChild(totalDiv);
}

window.runJsBenchmark = _runJsBenchmark;

File: Astro-TFG/public/scripts/6-patrones/prueba-1/version-1/python/main.py
import time
import json
from pyscript import PyWorker, display
import js  # type: ignore
import asyncio

worker = None
worker_time = 0


async def initialize_worker():
    global worker
    if worker is None:
        base = js.window.location.origin + js.window.document.body.dataset.pyPath
        worker = PyWorker(
            f"{base}worker.py",
            type="pyodide",
            config=f"{base.replace('python', 'json')}pyscript-worker.json"
        )
        await worker.ready


async def launch_worker(event):
    try:
        global worker, worker_time
        start_worker_time = time.perf_counter()
        await initialize_worker()
        worker_time = (time.perf_counter() - start_worker_time) * 1000
        json_str = await worker.sync.do_analisis()
        result = json.loads(json_str)

        display_result(result)
        js.window.hideExecutionLoader()
    except Exception as e:
        display(f"Error: {e}", target="pyscript-output")


def display_result(r):
    global worker_time
    display(f"Worker init time: {worker_time:.2f} ms",
            target="pyscript-output")
    display(
        f"Training time: {r['training_time_ms']:.2f} ms", target="pyscript-output")
    display(
        f"Inference time: {r['inference_time_ms']:.2f} ms", target="pyscript-output")
    display(
        f"Accuracy: {r['accuracy']:.2f} %",  target="pyscript-output")
    display(
        f"Model size: {r['model_size_mb']:.2f} MB", target="pyscript-output")
    display(
        f"Overall time: {r['overall_time_ms']:.2f} ms", target="pyscript-output")


def js_run_py_benchmark(event):
    js.clearCell('pyscript-output')
    asyncio.ensure_future(launch_worker(None))


js.window.run_py_benchmark = js_run_py_benchmark
